File: php/ohlcv_to_csv.php
<?php
// Download paginated OHLCV candles and write them to a CSV file.
// Usage: php ohlcv_to_csv.php [SYMBOL TIMEFRAME DAYS]
require __DIR__ . '/vendor/autoload.php';

$symbol = $argv[1] ?? 'BTC/USDT';
$timeframe = $argv[2] ?? '1h';
$days = (int) ($argv[3] ?? 7);

$exchange = new \ccxt\binance();
$cursor = $exchange->milliseconds() - $days * 86400 * 1000;
$candles = [];
while (true) {
    $batch = $exchange->fetch_ohlcv($symbol, $timeframe, $cursor, 1000);
    if (count($batch) === 0) {
        break;
    }
    $candles = array_merge($candles, $batch);
    $lastTimestamp = $batch[count($batch) - 1][0];
    if (count($batch) < 1000 || $lastTimestamp <= $cursor) {
        break; // reached the newest candle
    }
    $cursor = $lastTimestamp + 1;
}

$filename = str_replace(['/', ':'], '-', $symbol) . "-{$timeframe}.csv";
$file = fopen($filename, 'w');
fputcsv($file, ['timestamp', 'datetime', 'open', 'high', 'low', 'close', 'volume']);
foreach ($candles as $c) {
    fputcsv($file, [$c[0], $exchange->iso8601($c[0]), $c[1], $c[2], $c[3], $c[4], $c[5]]);
}
fclose($file);
echo count($candles), " candles -> {$filename}\n";

File: python/ohlcv_to_csv.py
# Download paginated OHLCV candles and write them to a CSV file.
# Usage: python ohlcv_to_csv.py [SYMBOL TIMEFRAME DAYS]
import csv
import sys
import ccxt

symbol = sys.argv[1] if len(sys.argv) > 1 else 'BTC/USDT'
timeframe = sys.argv[2] if len(sys.argv) > 2 else '1h'
days = int(sys.argv[3]) if len(sys.argv) > 3 else 7

exchange = ccxt.binance()
cursor = exchange.milliseconds() - days * 86400 * 1000
candles = []
while True:
    batch = exchange.fetch_ohlcv(symbol, timeframe, cursor, 1000)
    if not batch:
        break
    candles += batch
    last_timestamp = batch[-1][0]
    if len(batch) < 1000 or last_timestamp <= cursor:
        break  # reached the newest candle
    cursor = last_timestamp + 1

filename = f"{symbol.replace('/', '-').replace(':', '-')}-{timeframe}.csv"
with open(filename, 'w', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(['timestamp', 'datetime', 'open', 'high', 'low', 'close', 'volume'])
    for c in candles:
        writer.writerow([c[0], exchange.iso8601(c[0])] + c[1:])

print(f"{len(candles)} candles -> {filename}")
